keps/draft-20190516-KUDO-package-manager.md

```diff
@@ -0,0 +1,176 @@
+
+### Goals
+
+Specific goals are:
+* Coming up with a structure for a Package
+* Defining the overall `kubebuilder/framework` repository structure
+* Having a solution on how to distribute this structure
+* Having this platform-agnostic so Packages can be distributed across multiple solutions
+* Adapting concepts from other successfull open-source projects in this regard
+
+### Non-Goals
+
+Non-goals are:
+* Re-inventing the wheel
+* Solving problems that are not related to the overall goals
+* Some more that I can't think of yet
+
+## Proposal
+
+This is where we get down to the nitty gritty of what the proposal actually is. The folder structure and what could be interpreted as our v0 of the Registry was originally discussed in PR [#87](https://github.com/kudobuilder/kudo/pull/87) and is a great way to start when thinking of this enhancement.
+
+### User Stories
+
+#### Framework Developer
+
+As a Framework Developer I, ...
+
+* would like to be able to host validation/testing plans and resources as part of the package
+
+#### Cluster Administrator
+
+As a Cluster Administrator I, ...
+
+* would like to be able to host my custom Framework extensions on a private KUDO Framework repo
+* would like to be able to add additional validation/testing plans and resources
+
+### Implementation Details/Notes/Constraints
+
+We need to make sure that Packages can be easily accessed even if the underlying Storage Backend changed. Therefor, adding a middle layer like a registry server that has multiple Storage Backends as a swappable engine underneath seems like a good pattern. This also would enable users to keep track or restrict access to specific Frameworks.
+
+The idea would be a HTTP Server that can be easily accessed, even in air-gapped environments, depending on the user needs. As Storage Backends there could be a multitude such as `Local`, `Google`, `S3`, `Github`, `Minio`, `Docker` and so on.
+Having as an API a simple HTTP Server that just serves essentially a single yaml file that holds all the information also would enable us as a first iteration to use any HTTP server. This should work as it already with Google Cloud Storage or S3, all it requires is having access to the yaml file and something like the KUDO registry that serves like a middle layer inbetween could be easily added later on.
+
+Some caveats to this could for instance be defining a clean interface that won't break things when a user attempts to access Frameworks and hosted repos. Another caveat is deciding on the right proper structure in how a Package and its higher repo structure will look like. Design decisions here could potentialy impact future implementations of Storage Backends we haven't thought of yet. Another caveat would be identifying the right approach in versioning as this dictates a lot also how the structure will look like.
+
+The Repo structure on your local laptop could look in the short term as follow:
+
+```bash
+/index.yaml
+/kafka
+/kafka/2.2.0
```

Comment:
2.2.0 refers to the kafka version, correct? do we need another folder for the framework version (i.e., I update the framework but not the underlying kafka version)?'
Alternatively, we could have something like /kafka/2.2.0-0.1.0/kafka-framework.yaml

Reply:
Good point. `2.2.0` would be the SemVer of the KUDO Framework and **not** the corresponding base tech version (in this case Kafka). The base tech version would be defined within the `index.yaml`, e.g. like [here](https://github.com/fabianbaier/kudo/blob/cli-improvement/pkg/kudoctl/util/repo/repo.go#L21)

Reply:
👍 to what Fabian said